client/allocrunner/consul_hook.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/consul"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/client/widmgr"
+	"github.com/hashicorp/nomad/nomad/structs"
+	structsc "github.com/hashicorp/nomad/nomad/structs/config"
+)
+
+const (
+	// consulTokenFilePrefix is the begging of the name of the file holding the
+	// Consul SI token inside the task's secret directory. Full name of the file is
+	// always consulTokenFilePrefix_identityName
+	consulTokenFilePrefix = "nomad_consul"
+
+	// consulTokenFilePerms is the level of file permissions granted on the file in
+	// the secrets directory for the task
+	consulTokenFilePerms = 0440
+
+	// consulServicesAuthMethodName is the JWT auth method name that has to be
+	// configured in Consul in order to authenticate Nomad services.
+	consulServicesAuthMethodName = "nomad-workloads"
+
+	// consulTemplatesAuthMethodName the JWT auth method name that has to be
+	// configured in Consul in order to authenticate Nomad templates.
+	consulTemplatesAuthMethodName = "nomad-templates"
+)
+
+type consulHook struct {
+	alloc         *structs.Allocation
+	allocdir      *allocdir.AllocDir
+	widmgr        widmgr.IdentityManager
+	consulConfigs map[string]*structsc.ConsulConfig
+	hookResources *cstructs.AllocHookResources
+
+	logger log.Logger
+}
+
+func newConsulHook(logger log.Logger, alloc *structs.Allocation,
+	allocdir *allocdir.AllocDir,
+	widmgr widmgr.IdentityManager,
+	consulConfigs map[string]*structsc.ConsulConfig,
+	hookResources *cstructs.AllocHookResources,
+) *consulHook {
+	h := &consulHook{
+		alloc:         alloc,
+		allocdir:      allocdir,
+		widmgr:        widmgr,
+		consulConfigs: consulConfigs,
+		hookResources: hookResources,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*consulHook) Name() string {
+	return "consul"
+}
+
+func (h *consulHook) Prerun() error {
+	job := h.alloc.Job
+
+	if job == nil {
+		// something crazy happened
+		err := fmt.Errorf("alloc %v does not have a job", h.alloc.Name)
+		h.logger.Error(err.Error())
+		return err
+	}
+
+	mErr := multierror.Error{}
+	tokens := map[string]string{}
+
+	// get consul clients
+	clients, err := getConsulClients(h.consulConfigs, h.logger)
+	if err != nil {
+		return err
+	}
+
+	for _, tg := range job.TaskGroups {
+		for _, service := range tg.Services {
+			if err := h.prepareConsulTokens(service, clients); err != nil {
+				mErr.Errors = append(mErr.Errors, err)
+				continue
+			}
+		}
+		for _, task := range tg.Tasks {
+			for _, service := range task.Services {
+				if err := h.prepareConsulTokens(service, clients); err != nil {
+					mErr.Errors = append(mErr.Errors, err)
+					continue
+				}
+			}
+		}
+	}
+
+	// store the tokens in hookResources
+	h.hookResources.SetConsulTokens(tokens)
+
+	return mErr.ErrorOrNil()
+}
+
+func (h *consulHook) prepareConsulTokens(service *structs.Service, clients map[string]consul.Client) error {
```

Comment:
This doesn't return tokens, so the map of `tokens` in the caller is never populated.